cpp/src/utils/metrics/metric.hh

```diff
@@ -0,0 +1,32 @@
+#ifndef TS_CPP_UTILS_METRICS_METRIC_HH_
+#define TS_CPP_UTILS_METRICS_METRIC_HH_
+
+#include <string>
+#include <set>
+#include <map>
+
+#include "src/utils/metrics/units.hh"
+
+namespace torchserve {
+    enum MetricType {
+        COUNTER,
+        GAUGE,
+        HISTOGRAM
+    };
+
+    class IMetric {
+        public:
+        IMetric(const MetricType& type, const std::string& name, const std::string& unit,
+                const std::set<std::string>& dimension_names) :
+                type{type}, name{name}, unit{Units::GetUnitMapping(unit)}, dimension_names{dimension_names} {}
+        virtual void AddOrUpdate(const std::map<std::string, std::string>& dimension_values, const double& value) = 0;
+
+        protected:
+        const MetricType type;
+        const std::string name;
+        const std::string unit;
+        const std::set<std::string> dimension_names;
```

Comment:
I believe `std::set` will maintain its elements in sorted order (ascending by default) but may not be the same order as in the metrics configuration file. This will not matter for the backend metrics implementation in this PR since it does not make any assumptions about the order in which the dimension names or values are provided. `AddOrUpdate` method accepts a map consisting of keys as dimension names and values as dimension values.

Do we want the implementation to use the metric dimension order specified in the configuration file?

Reply:
The backend should also keep the same order defined in the metrics configuration b/c backend metrics needs to be injected into prometheus metrics by frontend.